### ISSUE TITLE
Count(distinct) relative things

### ISSUE THREAD

**Issue by githubmanticore** [nck-2/test-rep#124]
## Abstract 
Count(distinct) expresses exact number of distinct elements in the set. 
 
Another well-known term for this is 'cardinality', which expresses the 'power' of the set. If set contains many repeated values - that has low cardinality. Say, gender in the list of people - despite the size of the list is 'male' and 'female'. If most of values are unique, such set has high cardinality. Say, phone num in the same list of people. So, the same list has low cardinality by gender, but high cardinality by phone num. 
 
Count(distinct) looking very simple, but in reality it sometimes implies not so simple details. 
 
1. Easiest case for count(distinct) is using one solid plain index. Simple count(*) assumes the same as count(distinct id), and that is always just count of the documents in resultset, and that is by design. All necessary data is already in the index, no additinal RAM necessary to calculate it. 
 
For counting among columns with low cardinality you have to dedicate the table sized at most as value of cardinality. That is perfectly ok if you want count(distinct gender), as in that case table of 2 values is enough. 
 
Grey area is counting values for columns with unknown or high cardinality. Say, if you want to count (distinct phone) where id between 1 and 1000 - well, you need table with at most 999 cells. If you want to count (distinct phone.substr(0,3)) without limits - that is also ok and reasonable. But if you by mistake run query which will retrieve whole billion of phone numbers in order to calculate count(distinct) - you falls into problems, since managing such big array will consume many local resources (cpu, RAM and m.b. even disk). If just loosing of 1 cpu core for a hour is not a big deal, but being killed by OOM killer is more serious in the case. Especially when your indexes are mlocked and so, quantity of really free RAM is not enough. 
 
2. Counting among chunks is non trivial when keeping partitioning benefits. That is about _local_ chunks. Into this category falls both distributed made from locals, and RT indexes as is (taking in mind that their internal partitioning is not an exception, but exactly usual working state). 
 
Key benefit of partitioning is that all parts are independent, and so, may be processed in separate jobs scheduled anyway (either parallel, either sequental - no difference). 
 
Serial processing allows to use just one set of data for collecting aggregates and groups, just by process part by part with only one sorter, accumulating results. That is plain old way, and it perfectly works with all cardinality-depends tasks (among explicit 'count (distinct)' we also usually return words stats, which is not only implies 'count (distinct words)' over every query, but even provides details - the distinct words themselves with statistic. So, it is surely an example of cardinality processing also). Weak part of serial processing is that we use only one CPU core, since work serially. 
 
Parallel processing allows to utilize more CPU and in general finish whole task quicker. But it implies in addition another task: after all parts processed we have to aggregate all results into one. 
 
On huge tables with low cardinality parallel processing wins: on finish we quickly aggregate small (because of low cardinality) intermediate results and finish whole query faster then if we process parts one-by-one. 
 
On huge table with high cardinality parallel processing might be slower, then serial. Say, if we collect phone numbers, and each part provides distinctly different - well, we parallelly collect, say, 10 sets by 1G, and finally merge them into final 1 set of 10G. Comparing with serial work we just made useless intermediate steps, and final huge merge just ate out all achieved benefits. Well, at least nothing fatally wrong happens. 
 
On huge tables with unknown or medial cardinality parallel processing might be both slower, and more RAM-consumable. Imagine same 10 parts have distributed 1G of phone numbers, so that each part by itself has 1G of distinct numbers, and whole index also has same 1G. When processing in parallel, we, first, need 10G of RAM, And then also need time to merge this 10G. 
Finally we will have 1G of results, but we both spend many time for merging, and spend extra 9G of RAM for intermediate results. In this case simple serial processing might be much more efficient both by speed and RAM consumption. 
 
(We use combined approach: we have limited N of threads which processes parts in parallel, and each thread, in turn, processes available tasks from the queue one-by-one. Having 200 disk chunks and 4 parallel threads, each of them will sequentally process approx 50 parts, and finally we will merge 4 results. But 200 chunks is quite extremal size; usually rt index with 30-40 chunks on 8-cores CPU will be processed more 'parallel' (8 threads) then 'sequental' (5 parts in serie)) 
 
Main thought here is that high parallelization is not always give benefits, but sometimes leads to big overloading and resource consumption comparing to serial processing. And that is unpredictable (say, you can calculate cardinality of different static attributes, and make decision 'how to process' basing on that. But if it is dynamic attr with unknown value - it seems that safest will be throw out parallelism and use simple sequental processing.) 
 
3. Counting among remote nodes is even more complex. To calculate exact count we need to aggregate data in the one set or state. That is - if one remote calculated 1.5 billion, another - 1 billion, we can't count overall distincts from knowning these two numbers. The only thing we can say just knowing numbers without collecting data itself - that final count is something between 1.5 and 2.5 billions. But if we need exact value - we need literally fetch this 1.5 billion distincts from one node, 1 billion from second, and count distinct values from these collected sets. So, the task is very similar to collecting final aggregate from results of parallel processing parts, but make the things harder by the fact that these results have to be fetched over network to one node in order to be actually counted. 
 
## Facts 
 
- Count(distinct) works precisely only when processed by single sorter. That is: 
1. Over single plain index (always) 
2. Over combination of local parts in case of sequental processing (i.e. single sorter, no parallel threads) 
 
- count(distinct) returns non-precise value (better call it 'grade', not 'count') from several sorters. That is: 
1. From >1 remotes over network 
2. From any combination of local parts in case of parallel processing 
 
Local parts combination is either direct list of indexes (like select ... from index1,index2,index3), either distributed from locals, either rt-index with many chunks. For now 'grade' returns sum of results, which is mathematically high bound of possible count. 
 
- more parts in index - less precise grade. For just sum of part's counts it is sometimes definitely wrong, as 'count (distinct gender) where num_of_children_born_by_person>0' for 10 parts may say 10. 
 
- The way how multi-part query will be processed is unpredictable in general. If they will be processed parallel - result will be unprecise 'grade' with higher value. If sequental - result will be precise 'count'. All we can say for now - it is grade 'in general'. 
 
 
## What is seems to be reasonable to do 
 
- Describe the problem in the docs. That is not a bug, nor error. That is just a task which may require high resource consumption for precise calculation.  
 
- Question: How robust is to have precise high values? I.e., is it really necessary to return exactly '1000004' instead of just '<a href ="https://gitlab.com/manticoresearch/dev/labels/1000">~1000</a>000', or even '>500000'? 
 
- Hack: if you _need_ precise value over local multi-part index - use VIP connection. It is run in single dedicated thread. So, all processing will be serialized, and all values will be precise 
 
- It looks like in real life high-cardinality values are necessary to be counted precise only in very specific tasks (I can't even provide the one). In general it seems that, say, values like 5 and 7 need to be distinguished, but like 1005 and 1007 - probably no. So, for later answer as <a href ="https://gitlab.com/manticoresearch/dev/labels/1000">~1000</a> is enough, and no precise summing necessary. 
 
## Ideas 
 
- If we can know total cardinality, it would be good hint to be more precise in grades. Say, with 'gender' - doesn't matter, how many chunks in your index, you never will have this value >2 (well, I use traditional gender by born, not modern strange deviations). 
 
- N mirrors of the same index has same cardinality, and it not need to be summed on result.  
 
- In turn, if we split an index to parts, having values of some columns non-intersected among parts - in opposite, direct sum of results from all parts will give exact result.  
 
In last 2 cases we can return precise results despite of having parallel or remote processing. But the knowledge, whether it could work this way is not 'generally known' for us. So, if in such cases user can help - it will allow to return more precise result. Say, we can return row of numbers from parts - and he may say what is to do, like 'take max', or 'sum them'. 
And that is attribute of query, not the index itself (say, 1 partition may contains info about all people with age >60, second - all about younger. Having this knowledge when counting ages I can ask to 'sum' intermediate values and will surely get precise and true result. But if I want to count, say, by gender - then knowledge about age splitting is useless, and same 'summing' will give you up to 4, which is definitely wrong. In this case providing rule like 'max (Nval, 2)' where Nval is value from concrete where 2 is overall cardinality of genders will provide definitely more precise answer. So, to say shorter 
 
- Sometimes it sounds reasonable if user may provide formulae for count collection. 
 
- For low cardinality we can always provide precise result (since it is not very resource-greedy to calculate it in all cases). 
 
- For low cardinality even over remotes - that sounds reasonable to do all things, like literal sending of unique set of values to master, count them and return exact result. Say, that is no problem to do so if you have set of 50 values (but is not so obvious, say, if you have 50K of them). 
 
- We can tune edge value to switch from 'count' to 'grade'. 
Say, 10K of values, or 10Kb of dataset size (as example). That is - if agent produced up to limit distinct values for counting - it will send whole dataset, and master then will be able to produce precise 'count' value. 
If dataset is huge - well, agent will just return a number. And master will collect and return 'grade' instead of precise count. 
 
- Going further - if 9/10 agents returned whole datasets, and 10-th returned just a number - we could anyway return more precise grade. Counting precise cardinality of first 9 parts effectively make grade calculation as one from 2 parts (one calculated, and one returned), instead as one from 10 parts, which is less precise anyway. 
 
- Going back - if we know, that we're running on agent, and we know the dataset is huge, so that master will anyway perform kind of averaging over agents - then it may be reasonable NOT to count precise cardinality for ourselves, since it is not will be necessary 'so precise'. We may improve our own calculation, if possible. Say, if processing 10K docs gives 1K distincts, next 10K - 1K more, we may guess that processing whole 1M of docs will gives total ~100K distincts. So, let's stop collect and just multiply N of docs to 0.1 for such grade. 
 
- For local parallel processing - consider, if shared collector may work sometimes better. For now we have 'distinct counter' in sorter, where we collect grouped values and sometimes perform 'uniq', throwing out dead groups. That is isolated single-threaded construct. Say, if we may share such counter over all parallel sorters. Negative thing is it will be shared, and so, need to be guarded. Data race will make it hot place consuming time in locks. But in situation of 'worst' case - when 10 sorters produces 10 distincts by 1G which finally reduced to single 1G - such shared calculation might work. Collecting these 1G of cardinal values it yes, will be races. But when distincts saturated - the rest will just check already inserted values, and will not race, and will not consume additional 9G of RAM. So, it sounds like a viable idea to check. Another thought about the same - we can try to use kind of hash instead of vec, and even lockless. It consume a bit more RAM, but looks like may be optimized for better and faster processing, despite of even being shared among several sorters. Also sounds viable to try. 
 
- Since shared is not tried/tested - m.b. it have sense to make an option to select it explicitly as query option (if it is tried and found overall better or same - m.b. switch without disturbing params). 
 
- M.b. make query option to select whether query calculation should be sequental or parallel? 
 
- M.b. no option, but intention: if we see 'count (distinct)' in the query - automatically switch to sequental calculation. 
 
- Previous two, but with option: by default count (distinct) in serial, other queries - in parallel, but if user doesn't need very exact count(distinct) and having just a grade is ok for him, he may literally ask for parallel execution 
 
- If precise value is not necessary at all, but average grade is completely enough for smbd, may be invent or copy the way how direct function 'cardinality' works? If it doesn't need many RAM, works faster, etc. then precise counting - we may provide such 'happy meal', it would be faster and less greedy. Say, use not 'count (distinct attr)' but 'cardinality(attr)', or 'count (distinct attr) ... option counting=relaxed'.
